File: backend/app/services/booking_text.py
"""Booking-text builder for Leasy / NetXp (spec 012, updated in spec 014).

Single source of truth — mirrored by the frontend for live preview but the
authoritative version lives here. Reads display fields off the Event (name,
date) and trip-shape fields off the Fahrbericht (funker, guest_count).
"""

from decimal import Decimal
from uuid import UUID

from ..models import BarItem, Event, Fahrbericht


def _fmt(amount: Decimal) -> str:
    return f"{amount:.2f}"


def build_booking_text(
    bericht: Fahrbericht,
    event: Event,
    bar_catalog: dict[UUID, BarItem],
) -> str:
    kiosk_lines: list[str] = []
    kiosk_total = Decimal("0")
    for bid, qty in bericht.kiosk_tally.items():
        if not qty:
            continue
        bar = bar_catalog.get(bid)
        if not bar:
            continue
        total = Decimal(qty) * bar.kb
        kiosk_total += total
        kiosk_lines.append(f"  {bar.name} × {qty} = € {_fmt(total)}")

    crew_lines: list[str] = []
    crew_cost = Decimal("0")
    for bid, qty in bericht.crew_tally.items():
        if not qty:
            continue
        bar = bar_catalog.get(bid)
        if not bar:
            continue
        total = Decimal(qty) * bar.ek
        crew_cost += total
        crew_lines.append(f"  {bar.name} × {qty} = € {_fmt(total)}")

    expenses_total = sum((e.amount for e in bericht.expenses), Decimal("0"))
    bar_surcharge = kiosk_total + crew_cost
    soll = bericht.boarding_fee + bar_surcharge
    cash = bericht.cash_amount or Decimal("0")
    diff = cash - soll

    funker = bericht.funker.display_name if bericht.funker else "—"
    guest_count = bericht.guest_count if bericht.guest_count is not None else "?"

    parts = [
        "KASSENBUCH-EINGANG:",
        f"Datum: {event.start_at.date().isoformat()}",
        f"Veranstaltung: {event.name} ({guest_count} Gäste)",
        f"Funker*in: {funker}",
        "",
        "EINNAHMEN (8400 / Erlöse Kiosk):",
    ]
    if kiosk_lines:
        parts.extend(kiosk_lines)
    else:
        parts.append("  — keine Kiosk-Einnahmen —")
    parts.extend(
        [
            "",
            f"SUMME EINNAHMEN: € {_fmt(kiosk_total)}",
            f"Umlage Boarding: € {_fmt(bericht.boarding_fee)}",
            f"Umlage Bar: € {_fmt(bar_surcharge)}",
            "─────────────────────────────────",
            f"SOLL Umschlag: € {_fmt(soll)}",
            f"IST Umschlag:  € {_fmt(cash)}",
            f"Differenz:     € {'+' if diff >= 0 else ''}{_fmt(diff)}",
            "",
            "AUSGABEN (4220 / Wareneinsatz Crew):",
        ],
    )
    if crew_lines:
        parts.extend(crew_lines)
    else:
        parts.append("  — keine Crew-Verköstigung —")
    parts.append(f"Crew-Wareneinsatz: € {_fmt(crew_cost)}")
    if expenses_total > 0:
        parts.append(f"Sonstige Ausgaben Fahrt: € {_fmt(expenses_total)}")
    return "\n".join(parts)

File: frontend/src/utils/bookingText.js
/**
 * Client-side booking-text renderer (spec 012, updated in spec 014).
 *
 * Mirrors `backend/app/services/booking_text.py` for instant preview. Reads
 * display fields off the Event (name, start_at) and trip-shape fields off the
 * Fahrbericht (funker, guest_count). If the formats ever diverge, the backend
 * is authoritative.
 */

function fmt(n) {
  return Number(n || 0).toFixed(2)
}

function eventDateString(event) {
  if (!event?.start_at) return ''
  return String(event.start_at).slice(0, 10)
}

export function buildBookingText({ bericht, event, catalog }) {
  const kioskLines = []
  let kioskTotal = 0
  for (const [bid, qty] of Object.entries(bericht?.kiosk_tally || {})) {
    if (!qty) continue
    const bar = catalog?.[bid]
    if (!bar) continue
    const total = qty * Number(bar.kb)
    kioskTotal += total
    kioskLines.push(`  ${bar.name} × ${qty} = € ${fmt(total)}`)
  }

  const crewLines = []
  let crewCost = 0
  for (const [bid, qty] of Object.entries(bericht?.crew_tally || {})) {
    if (!qty) continue
    const bar = catalog?.[bid]
    if (!bar) continue
    const total = qty * Number(bar.ek)
    crewCost += total
    crewLines.push(`  ${bar.name} × ${qty} = € ${fmt(total)}`)
  }

  const expensesTotal = (bericht?.expenses || []).reduce(
    (s, e) => s + Number(e.amount || 0),
    0,
  )
  const boarding = Number(bericht?.boarding_fee || 0)
  const surcharge = kioskTotal + crewCost
  const soll = boarding + surcharge
  const cash = Number(bericht?.cash_amount || 0)
  const diff = cash - soll

  const funker = bericht?.funker?.display_name || '—'
  const guestCount = bericht?.guest_count ?? '?'

  const parts = [
    'KASSENBUCH-EINGANG:',
    `Datum: ${eventDateString(event)}`,
    `Veranstaltung: ${event?.name || ''} (${guestCount} Gäste)`,
    `Funker*in: ${funker}`,
    '',
    'EINNAHMEN (8400 / Erlöse Kiosk):',
  ]
  if (kioskLines.length) parts.push(...kioskLines)
  else parts.push('  — keine Kiosk-Einnahmen —')
  parts.push(
    '',
    `SUMME EINNAHMEN: € ${fmt(kioskTotal)}`,
    `Umlage Boarding: € ${fmt(boarding)}`,
    `Umlage Bar: € ${fmt(surcharge)}`,
    '─────────────────────────────────',
    `SOLL Umschlag: € ${fmt(soll)}`,
    `IST Umschlag:  € ${fmt(cash)}`,
    `Differenz:     € ${diff >= 0 ? '+' : ''}${fmt(diff)}`,
    '',
    'AUSGABEN (4220 / Wareneinsatz Crew):',
  )
  if (crewLines.length) parts.push(...crewLines)
  else parts.push('  — keine Crew-Verköstigung —')
  parts.push(`Crew-Wareneinsatz: € ${fmt(crewCost)}`)
  if (expensesTotal > 0) {
    parts.push(`Sonstige Ausgaben Fahrt: € ${fmt(expensesTotal)}`)
  }
  return parts.join('\n')
}
